components/composable/notebook.js

```diff
@@ -0,0 +1,399 @@
+// @flow
+import React from "react";
+
+import SyntaxHighlighter from "react-syntax-highlighter";
+import ReactMarkdown from "react-markdown";
+import { idea } from "react-syntax-highlighter/dist/styles";
+
+export const Passthrough = (props: { children?: React$Node }) =>
+  props.children || null;
+
+export const Output = () => <pre>I am output</pre>;
+
+type OutputsProps = {
+  children: React$Node,
+  className?: string
+};
+
+class Outputs extends React.Component<OutputsProps> {
+  static defaultProps = {
+    children: null,
+    className: "outputs"
+  };
+
+  render() {
+    if (this.props.children) {
+      return <div className={this.props.className}>{this.props.children}</div>;
+    }
+    return null;
+  }
+}
+
+type PromptProps = {
+  counter?: number | null,
+  running?: boolean,
+  queued?: boolean
+};
+
+// Totally fake component for Markdown consistency
+class PromptBuffer extends React.Component<*> {
+  render() {
+    return null;
+  }
+}
+
+class Prompt extends React.Component<PromptProps> {
+  static defaultProps = {
+    counter: null,
+    running: false,
+    queued: false
+  };
+
+  text = () => {
+    if (this.props.running) {
+      return "[*]";
+    }
+    if (this.props.queued) {
+      return "[…]";
+    }
+    if (this.props.counter) {
+      return `[${this.props.counter}]`;
+    }
+    return "[ ]";
+  };
+
+  render() {
+    return this.text();
+  }
+}
+
+type EditorProps = {
+  language: string,
+  children: string | React$Element<any>,
+  className?: string
+};
+
+class Editor extends React.Component<EditorProps> {
+  static defaultProps = {
+    children: "",
+    language: "python",
+    className: "input"
+  };
+
+  render() {
+    if (typeof this.props.children === "string") {
+      return (
+        <SyntaxHighlighter
+          style={idea}
+          language={this.props.language}
+          className={this.props.className}
+          customStyle={{
+            padding: "10px 0px 10px 10px",
+            margin: "0px",
+            backgroundColor: "var(--cm-background)"
+          }}
+        >
+          {this.props.children}
+        </SyntaxHighlighter>
+      );
+    }
+    return this.props.children;
+  }
+}
+
+type InputProps = {
+  children: React$Node
+};
+
+class Input extends React.Component<InputProps> {
+  render() {
+    const children = React.Children.map(this.props.children, child => {
+      if (child.type === Prompt || child.type === PromptBuffer) {
+        return <div className="prompt">{child}</div>;
+      } else if (child.type === Editor) {
+        return (
+          <div className="input">
+            {React.cloneElement(child, { className: "input" })}
```

Comment:
Why are you passing the `className` prop to the `Editor` element if it is already set as the default?

Reply:
I was running into CSS issues, looks like I can delete the wrapper div now and just set it on the `child` (which was necessary for the `ReactHighlighter` to look right).